src/statement.ts

```diff
@@ -0,0 +1,240 @@
+    opts?: Signal
+  ): Promise<Result<T[K]>>;
+  async all<T = S, K extends keyof T = keyof T>(
+    colName?: K,
+    opts: Signal = {}
+  ): Promise<Result<T | T[K]>> {
+    try {
+      const start = performance.now();
+      const { sql, type, tables } = await this.#parseAndExtract();
+      switch (type) {
+        case "read": {
+          const config = await extractReadonly(this.config, {
+            type,
+            tables,
+          });
+          const results = await queryAll<T>(config, sql, opts);
+          if (colName != null) {
+            return wrapResult(
+              extractColumn(results, colName),
+              performance.now() - start
+            );
+          }
+          return wrapResult(results, performance.now() - start);
+        }
+        default: {
+          let receipt = await exec(this.config, { type, sql, tables });
+          if (this.config.autoWait ?? false) {
+            const waited = await receipt.wait();
+            receipt = { ...receipt, ...waited };
+          }
+
+          return wrapResult(receipt, performance.now() - start);
+        }
+      }
+    } catch (cause) {
+      if (
+        cause instanceof Error &&
+        cause.message.includes("column not found")
+      ) {
+        throw new Error("COLUMN_NOTFOUND", { cause });
+      }
+      throw new Error("ALL_ERROR", { cause });
+    }
+  }
+
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  async first<T = S, K extends keyof T = keyof T>(): Promise<T>;
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  async first<T = S, K extends keyof T = keyof T>(
+    colName: undefined,
+    opts?: Signal
+  ): Promise<T>;
+  async first<T = S, K extends keyof T = keyof T>(
+    colName: K,
+    opts?: Signal
+  ): Promise<T[K] | null>;
+  async first<T = S, K extends keyof T = keyof T>(
+    colName?: K,
+    opts: Signal = {}
+  ): Promise<T | T[K] | null> {
+    try {
+      const { sql, type, tables } = await this.#parseAndExtract();
+      switch (type) {
+        case "read": {
+          const config = await extractReadonly(this.config, {
+            type,
+            tables,
+          });
+          const results = await queryFirst<T>(config, sql, opts);
+          if (results == null || colName == null) {
+            return results;
+          }
+          return extractColumn(results, colName);
+        }
+        default: {
+          const receipt = await exec(this.config, { type, sql, tables });
+          /* c8 ignore next */
+          if (this.config.autoWait ?? false) {
+            await receipt.wait();
+          }
+          return null;
+        }
+      }
+    } catch (cause) {
+      if (
+        cause instanceof Error &&
+        cause.message.includes("column not found")
+      ) {
+        throw new Error("COLUMN_NOTFOUND", { cause });
+      }
+      throw new Error("FIRST_ERROR", { cause });
+    }
+  }
+
+  async run(opts: Signal = {}): Promise<Result<never>> {
```

Comment:
Essentially, only return metadata.

Reply:
This name feels counter intuitive to me, e.g. I run a query to see results, but I'm guessing this is how the D1 api works?

Reply:
Yeh, run is the side-effects version, comes from D1, better-sqlite3, etc